### ISSUE TITLE
Vente privée seems broken

### ISSUE THREAD

**Issue by clochix** [cozy-labs/konnectors#351]
A user reported (by email) that he was unable to read some of the PDFs downloaded the the vente-privée site.
Looks like this files are in fact HTML error pages, often requesting to log into the site.

So something probably goes wrong with auth or cookies. Password is ok, the connector is able to fetch the list of bills and their metadatas, but not the attached PDF file.

**Comment by clochix** [comment]
Pinging @SomeAverageDev

**Comment by SomeAverageDev** [comment]
Hello,
Only the first bill is fetched for me, but the parsing of the VP page is fine.
The issue might be located in the fetching process inside konnector itself...

How can we debug that part ?

**Comment by ZeHiro** [comment]
@SomeAverageDev, can't you ise the option jar=true of request for saving the cookies?

**Comment by clochix** [comment]
This issue is really strange, the parsing of the VP page is fine but the downloading of the bill itself triggers auth error.

Maybe could you run the konnector locally, in DEBUG mode to have more logs. Or add debug output.

**Comment by SomeAverageDev** [comment]
This might be VP that changed something for security reason... they might think no human would download all its bills within 2 seconds.

On my dev, I run the cozy-dev vagrant image, do you know how I can enable the debug on konnectors side ?
On my vp konnector I use the "`[debug](https://www.npmjs.com/package/debug)`" library. This prints the result of the parsing (url / dates / prices), which are good for me.

**Comment by ZeHiro** [comment]
For debugging I use console.log, To insert a delay between imports, this can be done using setTimeout(delay, callback) : 
here : 
https://github.com/cozy-labs/konnectors/blob/master/server/lib/save_data_and_file.coffee#L64

and define a new option. I would see something like that : 

``` coffeescript
if options.downloadTimeout? and typeof (options.downloadTimeout) is 'number'
    setTimeout options.downloadTimeout, createFileAndSaveData(entry, entryLabel)
else
    createFileAndSaveData entry, entryLabel
````

@frankrousseau ?

**Comment by nicofrand** [comment]
I just tried this konnector today (2016-07-13) and here is what happened, maybe that could help:

* It retrieved all my bills
* Several have the same weird size : 796B
* I can open only the two last bills (even the others bills with sizes around 10 kB do not work)

![4](https://cloud.githubusercontent.com/assets/3419050/16794942/eff12412-48da-11e6-8866-c7f6b7995651.png)

**Comment by ZeHiro** [comment]
@nicofrand 
The small 'bills' must be html files. Their content could help to understand what's going on.

**Comment by nicofrand** [comment]
I downloaded one the 796B files.
Once downloaded tis size is around 11Kio (like the two others between 10 and 76 kB in fact), and not 796B !

Here is the content :

```html


<!DOCTYPE html>
<html lang="fr-FR" 
>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />


    
        <meta name="description" content="" />
        <title>Vente-privee.com : grandes marques à prix discount, ventes privées chaque jour. </title>



    

        <link rel="stylesheet" type="text/css" href="https://vp-eu.scene7.com/is/content/vpeu/inte/theme/20160708/00_20160708_fr_hme.css">
    <link rel="stylesheet" type="text/css" href="https://static.vente-privee.com/vp5/_web/site/vp/_common/css/gabarit.css">
    <link rel="stylesheet" type="text/css" href="https://static.vente-privee.com/vp5/_web/site/vp/_common/css/jplayer.vp.skin.css">
    <link rel="stylesheet" type="text/css" href="https://static.vente-privee.com/vp5/_web/site/vp/_common/css/login.css">

    <script type="text/javascript">
        window.nocookie = true;
        if (document.documentMode === 7) {
            document.write('<link rel="stylesheet" type="text/css" href="https://static.vente-privee.com/vp5/_web/site/vp/_common/css/styles-ie.css">');
            document.write('<script src="https://static.vente-privee.com/vp5/_web/site/vp/_common/js/html5.js"></' + 'script>');
        }
        if (document.documentMode === 8) {
            document.write('<script src="https://static.vente-privee.com/vp5/_web/site/vp/_common/js/html5.js"></' + 'script>');
        }
    </script>

</head>
<body class="fr login">
    <div id="wrapper">
        <div id="container">
            <!-- BEGIN: Header Desktop -->

            

<div id="headerWrapper" class="showDesk headerBlock">
    <div class="cookiesBloc" style="display:none;">
        <p>En poursuivant votre navigation sur ce site, vous acceptez l'utilisation de cookies pour vous proposer des services et offres adapt&eacute;s &agrave; vos centres d'int&eacute;r&ecirc;ts. </p>
        <p>Pour en savoir plus et param&eacute;trer les cookies, consultez <a onclick="window.open(this.href, 'dataProtection', 'height=440, width=680, left=50, top=50, status=no, toolbar=no, menubar=no, location=no, scrollbars=yes');return false;" href="/registration/PrivacyPolicy#pp_cookies">notre Politique de Confidentialit&eacute; et Protection de la Vie Priv&eacute;e</a></p>
        <a class="cookiesClose" href="">Fermer</a>
    </div>
    <header role="banner">
        <h1><a href="/vp4/Login/Portal.ashx" class="xitiev" data-xiti="header::hors_log::logo_accueil" data-clicktype="N" title=""></a></h1>
                
            <div class="linksRight">
                                    <a id="nav_registration" class="inscBtn xitiev" data-xiti="header::hors_log::devenir_membre" data-clicktype="N" href="https://secure.fr.vente-privee.com/registration/" title="">Devenir membre</a>
            </div>
            </header>
    </div>



            <!-- END: Header Desktop -->
            <header class="showMob">

                <h1>vente-privee.com</h1>
            </header>

            <div id="content" role="main">
                <!-- START: MainCol -->
                <div id="mainCol">
                    



<script>
    var viewmodelparam = {
        validationonerror: 'False',
        operationForm: false,
        isVoyage: 'False',
        registrationUrl: '/registration/registration?CountryCode=FR&'

    };
</script>




    <div class="chooseLanguage">
        <h3>Choisissez votre pays :</h3>
        <ul class="listCountry"><li class="default"><a href="javascript:void(0)" id="frLoginLink" title="France">France</a></li><li class="country"><ul><li><a href="https://secure.de.vente-privee.com/authentication/login/DE" id="deLoginLink" title="Allemagne">Allemagne</a></li><li><a href="https://secure.es.vente-privee.com/authentication/Login/ES" id="esLoginLink" title="Espagne">Espagne</a></li><li><a href="https://secure.it.vente-privee.com/authentication/Login/IT" id="itLoginLink" title="Italie">Italie</a></li><li><a href="https://secure.uk.vente-privee.com/authentication/Login/EN" id="enLoginLink" title="Royaume-Uni">Royaume-Uni</a></li><li><a href="https://secure.fr.vente-privee.com/authentication/login/BE" id="beLoginLink" title="Belgique">Belgique</a></li><li><a href="https://secure.de.vente-privee.com/authentication/login/AT" id="atLoginLink" title="Autriche">Autriche</a></li><li><a href="https://secure.eu.vente-privee.com/authentication/Login/NL" id="nlLoginLink" title="Pays-Bas">Pays-Bas</a></
li><li class="last"><a href="https://secure.us.venteprivee.com/vp4/Login/Login.aspx" id="usPortalLink" title="United States">United States</a></li></ul></li></ul>
    </div>
<div class="mdpModal" style="display: none;">
    <a class="closeModal" href="#" title="">&nbsp;</a>
    <h3>M&eacute;moriser mes identifiants</h3>
    <div class="mdpText">
        <p>Si vous souhaitez acc&eacute;der directement au site sans saisir votre mot de passe, activez simplement la case "m&eacute;moriser mes identifiants".</p>
<p>Si vous souhaitez modifier votre choix, supprimez les cookies enregistr&eacute;s dans votre navigateur.</p>
<p>Nous vous recommandons de ne pas activer cette fonction si vous utilisez un ordinateur public.</p>
    </div>
</div>
<div id="stopGrey" style="display: none;"></div>
<section>
    <h1>Déjà membre ?</h1>
    <h2>Identifiez-vous</h2>
    <div class="obj_form">
<form action="/authentication/login/FR?ReturnUrl=%2fmemberaccount%2forder%2finvoice%3forderId%3d144539546&amp;orderId=144539546" id="authenticationForm" method="post" name="authenticationForm">            <fieldset>
                <div class="formRow">
                    <label id="lbMail" for="mail">Votre email
                        <span>*</span>
                    </label>
                    <input data-val="true" data-val-mailvalidator="" data-val-mailvalidator-mailerror="Votre email et/ou votre mot de passe sont incorrects" data-val-mailvalidator-regex="^\w+(\.?[-+\w])*@\w+([-.]\w+)*\.[a-zA-Z]{2,80}$" id="mail" name="Mail" type="text" value="" />
                    <span id="aMailValidate" class="invalidLabel" title="" style="display: none;">
                        <span class="invalidBble showDesk">&nbsp;</span>
                        <span class="field-validation-valid invalidTxt" data-valmsg-for="Mail" data-valmsg-replace="true"></span>
                    </span>
                    <input type="hidden" id="FacebookId" name="FacebookId" value="" />
                    <input type="hidden" id="ExplicitFacebookConnect" name="ExplicitFacebookConnect" value="false" />
                    <input type="hidden" id="FacebookToken" name="FacebookToken" value="" />
                    <input type="hidden" id="FacebookConnectTries" name="FacebookConnectTries" value="0" />

                </div>

                <div class="formRow">
                    <label for="mdp" class="">Mot de passe
                        <span>*</span>
                    </label>
                    <input id="mdp" name="Password" type="password" />

                </div>

                
                            </fieldset>
            <div class="btWrapper">
                <span>
                    <input id="loginBtn" type="submit" value="Valider" class="primaryBt"/>
                </span>
            </div>
</form>    </div>
    <p class="lostPass">
        <a id="lostPsw" href="/authentication/Login/RecoverPassword">J'ai perdu mon mot de passe</a>
    </p>
    <div class="blocSubscribe">
        <h1>Devenez membre de vente-privee.com</h1>
        <p>Et bénéficiez de ses privilèges</p>
        <div class="btWrapper">
            <a id="SubsBtn" class="primaryBt" href="/registration/registration?CountryCode=FR&amp;">S'inscrire</a>
        </div>
    </div>
</section>
<footer>
    <p>Si vous rencontrez des problèmes pour vous connecter sur le site vente-privee.com, nous vous conseillons de fermer toutes vos fenêtres de navigateur Internet. Si le problème persiste, vous pouvez <a id="clearCookie" href="/vp4/registration/ClearCookie.aspx" >nettoyer les cookies de votre navigateur. </a><br />
Consulter notre<a id="privacyPol" onclick="window.open('/registration/PrivacyPolicy?CountryCode=FR', 'privacypol', 'height=430, width=648, left=50, top=50, status=no, toolbar=no, menubar=no, location=no, scrollbars=yes'); return false;" title="privacy_policy" href=""> Politique de Confidentialit&eacute; et Protection de la Vie Priv&eacute;e</a></p>
</footer>
<ul class="footerLinks">
    <li><a title="" target="_blank" class="tuvFooterLink" href="" onclick="window.open('http://secure.fr.vente-privee.com', 'dataProtection', 'height=480, width=720, left=50, top=50, status=no, toolbar=no, menubar=no, location=no, scrollbars=yes');return false;"></a></li>
    <li><span>&nbsp;</span><a target="_blank" href="" onclick="window.open('http://secure.fr.vente-privee.com', 'dataProtection', 'height=480, width=720, left=50, top=50, status=no, toolbar=no, menubar=no, location=no, scrollbars=yes');return false;" title=""></a></li>
</ul>
<span id="tagcriteo" style="display: inline-block; text-indent: -10000px">
    
</span>


                </div>
                <div class="clear"></div>
            </div>
        </div>
    </div>
    
    <script src="/authentication/bundles/jquery?v=gajZTh3bI9-vTEBUtjIq6EtzDQYIBGZlWDm55ARLX3Y1"></script>

    <script src="/authentication/bundles/cookiescnil?v=4uVQR3gaaZ7BE9JrwXYFyUkQYqvxR7yh5D8kCJqQoUQ1"></script>

    <script src="/authentication/bundles/jqueryval?v=mRjM0qa6T8GTCa8lhmXMI_-t5fsTCmHSxo4BqkY9x4A1"></script>

    <script src="/authentication/bundles/mailvalidation?v=t3JfIzdZvz30uuB0CoHM6hUO0t_qtWQZZ6MJJlL0-ws1"></script>



    <script type="text/javascript" src="https://static.vente-privee.com/vp5/_web/site/vp/_common/js/jquery.jplayer.min.js"></script>
    <script type="text/javascript" src="https://static.vente-privee.com/vp5/_web/site/vp/_common/js/gen.1.0.js"></script>
    
    <script type="text/javascript" src="https://static.vente-privee.com/vp5/_web/site/vp/_common/js/reg_customForm.js"></script>
    <script type="text/javascript" src="https://static.vente-privee.com/vp5/_web/site/vp/_common/js/login.js"></script>

    

    
<script type="text/javascript">
   
    xtnv = document;
    xtsd = "https://logs198";
    xtsite = "337001";
    xtn2 = "1";
    xtpage = "Page_de_Login";
    xt_multc = "";
    vp_xt_cookie = "secure.fr.vente-privee.com";
    xtdi = "";
    xt_an = "";
    xt_ac = "";
    xtergo = "0";
    if (window.xtparam != null) { window.xtparam += "&ac=" + xt_ac + "&an=" + xt_an + xt_multc; }
    else { window.xtparam = "&ac=" + xt_ac + "&an=" + xt_an + xt_multc; };
</script>
<script src="/authentication/bundles/xiti?v=nuAk3qq5IAjHGWo326fgHWiUdazJ8uOBK51QhK1e9XA1"></script>

<noscript>
    <img width="1" height="1" src='https://logs198.xiti.com/hit.xiti?s=337001&amp;amp;s2=1&amp;amp;p=Page_de_Login&amp;amp;di=&amp;amp;' alt='XiTi' />
</noscript>
    <script src='/authentication/WebAnalytics.axd' type="text/javascript"></script>

<script src="//cdn.optimizely.com/js/4620944294.js"></script>    
</body>
</html>
```

**Comment by ZeHiro** [comment]
@SomeAverageDev 
Maybe using the "keep me connected" in the connection form could fix the problem.

**Comment by SomeAverageDev** [comment]
thanks, we can see that while retreiving the bill file, the VP website redirect us to the login page.
and the login form might redirect back to the bill.

`<form action="/authentication/login/FR?ReturnUrl=%2fmemberaccount%2forder%2finvoice%3forderId%3d4654654&amp;orderId=14456446"`

We know the WHAT, but I don't understand the WHY yet. I think VP website updated something.
I need to try to set a delay between downloads.
In the meanwhile I made a version which downloads only the latest bill,  it could solve temporarly the issue...

**Comment by ZeHiro** [comment]
the "keep me connected" option wouldn't be enough @SomeAverageDev ?

**Comment by SomeAverageDev** [comment]
I will try, but usually this option is to keep connected through several days, not during the same session...
this could be an issue with cookies, security measures from VP, etc...
I managed to get the HTTP traffic yesturday, but could not decode the SSL yet...

quite difficult to find the root cause then.

**Comment by SomeAverageDev** [comment]
@ZeHiro 
the option "keep me connected" does not exist...

this issue is very strange, i have done the test just now and all 12 PDF were downloaded correctly.

Does everyone use a hosted cozy from cozy cloud ? It could be an IP filtering from VP seeing to much connection from the same IP... ?

**Comment by SomeAverageDev** [comment]
could someone having the issue can change a file on its cozy ?

replace in the file
`/usr/local/cozy/apps/konnectors/build/server/lib/save_data_and_file.js`
the line 60
`return createFileAndSaveData(entry, entryLabel);`
with
`return setTimeout (createFileAndSaveData, 3000, entry, entryLabel);`

**Comment by ZeHiro** [comment]
@SomeAverageDev 
I cannot test, I don't have a ventre privée account. 
You're fix looks like what I proposed a few comments ago, but my solution is more generic: it activates the delay only if the connector developper asks for it, so that it is not necessary to wait 3s for all the konnectors.

**Comment by nicofrand** [comment]
I tested your fix @SomeAverageDev and only 1 file was corrupted (501 B this time). So I guess that's better but that's not optimal. We could still increase the delay but that's an ugly hack :(.

**Comment by SomeAverageDev** [comment]
thanks @nicofrand, I don't really know what to do now...
Do you use an instance hosted by cozy cloud, or do you use your own ?
Another solution might be to configure a http proxy, but could raise security concerns...

I might try to send a mail to VP to see if they can give me some intels...

**Comment by SomeAverageDev** [comment]
@ZeHiro, i just wanted to try your solution before even thinking of asking for a state of the art code change ! ;-)

**Comment by ZeHiro** [comment]
@SomeAverageDev 
When you add this peace of code, could you also include the option in the wiki? Thanks

**Comment by nicofrand** [comment]
@SomeAverageDev I use my own instance on a kimsufi server.

**Comment by SomeAverageDev** [comment]
J'ai contacté vente-privée, même si j'ai aucune chance d'avoir une réponse ;-)

**Comment by SomeAverageDev** [comment]
a new release is avalaible, the new konnector only download the latest bill...only solution so far...

**Comment by nicofrand** [comment]
@SomeAverageDev Have you received any response from Vente-privée ?

**Comment by SomeAverageDev** [comment]
hell no !
I was not very optimistic on this one ;-)

**Comment by SomeAverageDev** [comment]
this makes me think that there should be a control on the konnector side, while we submit a PDF for downloading a bill, that the file downloaded is indeed a PDF (and not a HTML/error).

**Comment by nicofrand** [comment]
@SomeAverageDev I just tried to fetch my bills, and the connector only downloaded the last one, but even this only bill was corrupted. Maybe we could restore the previous behaviour which sometimes worked :D ?

**Comment by nicofrand** [comment]
Actually not sure the connector works at all.
Here is what I receive:

```html

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" > 

<html>
  <head>
    <title>PDFInvoice_WS</title>
    <meta name="GENERATOR" Content="Microsoft Visual Studio .NET 7.1">
    <meta name="CODE_LANGUAGE" Content="C#">
    <meta name=vs_defaultClientScript content="JavaScript">
    <meta name=vs_targetSchema content="http://schemas.microsoft.com/intellisense/ie5">
  </head>
  <body MS_POSITIONING="GridLayout">
	
    <form name="Form1" method="post" action="PDFInvoice_WS.aspx?i=8%7cn%7c5%7cG%7cS%7cG%7cK%7cu%7cm%7cL%7cz%7cx%7c5%7cB%7cU%7c5%7cY%7cu%7cf%7cZ%7c7%7cM%7cb%7cF%7cC%7cS%7cR%7cr%7cr%7cW%7cI%7cI%7c" id="Form1">
<input type="hidden" name="__VIEWSTATE" id="__VIEWSTATE" value="/wEPDwUKLTUxMTcwNzgxMGRk8GhrI2Odhe7sTbWoLyWp2T2w6eQ=" />


     </form>
	
  </body>
</html>
```

**Comment by SomeAverageDev** [comment]
it looks like they changed something...
I will check
there really need to have a document/mime type check on konnector while it downloads a document from an URL.

**Comment by SomeAverageDev** [comment]
I really don't understand, the URL is good, but with konnector, there is a redirection to the login form...

**Comment by nicofrand** [comment]
Yep, the user-agent seems ok though

**Comment by SomeAverageDev** [comment]
Ok I found something,
when the konnector works (usually the first time), the headers are as below

`            "search": "?orderId=12770000",
            "query": "orderId=12770000",
            "pathname": "/memberaccount/order/invoice",
            "path": "/memberaccount/order/invoice?orderId=12770000",
            "href": "https://secure.fr.vente-privee.com/memberaccount/order/invoice?orderId=12770000"
`

But when it dones not work (in a loop or after a first konnector execution)

`            "search": "?ReturnUrl=%2fmemberaccount%2forder%2finvoice%3forderId%3d12770000&orderId=127700001",
            "query": "ReturnUrl=%2fmemberaccount%2forder%2finvoice%3forderId%3d12770000&orderId=127700001",
            "pathname": "/authentication/login/FR",
            "path": "/authentication/login/FR?ReturnUrl=%2fmemberaccount%2forder%2finvoice%3forderId%3d12770000&orderId=127700001",
            "href": "https://secure.fr.vente-privee.com/authentication/login/FR?ReturnUrl=%2fmemberaccount%2forder%2finvoice%3forderId%3d12770000&ord
erId=127700001"
`
We can see that the URL built from the previous request is kind of "looping", and the encoding seems wrong...
It looks like konnector uses the same request object from one URL to another, which modifies the request "environnement" (headers)

**Comment by ZeHiro** [comment]
Theres is an error here in the konnector : 

https://github.com/cozy-labs/konnectors/blob/master/server/konnectors/vente_privee.js#L241

The connector should not `return next()`
Same here : 
https://github.com/cozy-labs/konnectors/blob/master/server/konnectors/vente_privee.js#L233

Calling next() will go to the next step in the table of functions (before all the bills are fetched) , and at the end go to 'loGout'. I thing the problem is here @nicofrand @SomeAverageDev

**Comment by SomeAverageDev** [comment]
looks like it's the issue...

**Comment by ZeHiro** [comment]
Closed by #508
